WalletWasabi/WabiSabi/Client/AmountDecomposer.cs

```diff
@@ -13,58 +16,67 @@ public class AmountDecomposer
 {
 	/// <param name="feeRate">Bitcoin network fee rate the coinjoin is targeting.</param>
 	/// <param name="allowedOutputAmount">Range of output amount that's allowed to be registered.</param>
-	/// <param name="outputSize">Size of an output.</param>
-	/// <param name="inputSize">Size of an input.</param>
 	/// <param name="availableVsize">Available virtual size for outputs.</param>
 	/// <param name="random">Allows testing by setting a seed value for the random number generator. Use <c>null</c> in production code.</param>
-	public AmountDecomposer(FeeRate feeRate, MoneyRange allowedOutputAmount, int outputSize, int inputSize, int availableVsize, Random? random = null)
+	public AmountDecomposer(FeeRate feeRate, MoneyRange allowedOutputAmount, int availableVsize, bool isTaprootAllowed, Random? random = null)
 	{
 		FeeRate = feeRate;
 
-		OutputSize = outputSize;
-
-		InputFee = FeeRate.GetFee(inputSize);
-		OutputFee = FeeRate.GetFee(outputSize);
-
 		AvailableVsize = availableVsize;
-
-		MinAllowedOutputAmountPlusFee = allowedOutputAmount.Min + OutputFee;
+		IsTaprootAllowed = isTaprootAllowed;
+		MinAllowedOutputAmount = allowedOutputAmount.Min;
 		MaxAllowedOutputAmount = allowedOutputAmount.Max;
 
 		Random = random ?? Random.Shared;
 
 		// Create many standard denominations.
-		DenominationsPlusFees = CreateDenominationsPlusFees();
+		Denominations = CreateDenominations();
+
+		ChangeScriptType = GetNextScriptType();
 	}
 
 	public FeeRate FeeRate { get; }
 	public int AvailableVsize { get; }
-	public Money MinAllowedOutputAmountPlusFee { get; }
+	public bool IsTaprootAllowed { get; }
+	public Money MinAllowedOutputAmount { get; }
 	public Money MaxAllowedOutputAmount { get; }
 
-	public Money OutputFee { get; }
-	public Money InputFee { get; }
-	public int OutputSize { get; }
-	public IOrderedEnumerable<ulong> DenominationsPlusFees { get; }
+	public IOrderedEnumerable<Output> Denominations { get; }
+	public ScriptType ChangeScriptType { get; }
+	public Money ChangeFee => FeeRate.GetFee(ChangeScriptType.EstimateOutputVsize());
 	private Random Random { get; }
 
-	private IOrderedEnumerable<ulong> CreateDenominationsPlusFees()
+	private ScriptType GetNextScriptType()
+	{
+		if (!IsTaprootAllowed)
+		{
+			return ScriptType.P2WPKH;
+		}
+
+		return Random.NextDouble() < 0.5 ? ScriptType.P2WPKH : ScriptType.Taproot;
+	}
+
+	private IOrderedEnumerable<Output> CreateDenominations()
 	{
-		ulong maxSatoshis = (ulong)MaxAllowedOutputAmount.Satoshi;
-		ulong minSatoshis = MinAllowedOutputAmountPlusFee;
-		var denominations = new HashSet<ulong>();
+		var denominations = new HashSet<Output>();
+
+		Output CreateDenom(double sats)
+		{
+			var scriptType = GetNextScriptType();
+			return Output.FromDenomination(Money.Satoshis((ulong)sats), scriptType, FeeRate);
```

Comment:
It would be good to understand what happens with duplicates here. It seems to me that 0, 1 and 2 should be the only duplications and they will removed later so, no problem but I don't remember if there are others.

Reply:
Added distinct